lib/clib/CCorresLemmas.thy

```diff
@@ -1040,41 +1047,17 @@ lemma ccorres_Guard_True_Seq:
    \<Longrightarrow> ccorres_underlying sr \<Gamma> r xf arrel axf A C hs a (Guard F \<lbrace>True\<rbrace> c ;; d)"
    by (simp, ccorres_rewrite, assumption)
 
-lemma ccorres_While_Normal_helper:
-  assumes setter_inv:
-    "\<Gamma> \<turnstile> {s'. \<exists>rv s. G rv s s'} setter {s'. \<exists>rv s. G rv s s' \<and> (cond_xf s' \<noteq> 0 \<longrightarrow> Cnd rv s s')}"
-  assumes body_inv: "\<Gamma> \<turnstile> {s'. \<exists>rv s. G rv s s' \<and> Cnd rv s s'} B {s'. \<exists>rv s. G rv s s'}"
-  shows "\<Gamma> \<turnstile> ({s'. \<exists>rv s. G rv s s' \<and> (cond_xf s' \<noteq> 0 \<longrightarrow> Cnd rv s s')})
-             While {s'. cond_xf s' \<noteq> 0} (Seq B setter)
-             {s'. \<exists>rv s. G rv s s'}"
-  apply (insert assms)
-  apply (rule hoare_complete)
-  apply (simp add: cvalid_def HoarePartialDef.valid_def)
-  apply (intro allI)
-  apply (rename_tac xstate xstate')
-  apply (rule impI)
-  apply (case_tac "\<not> isNormal xstate")
-   apply fastforce
-  apply (simp add: isNormal_def)
-  apply (elim exE)
-  apply (simp add: image_def)
-  apply (erule exec_While_final_inv''[where C="{s'. cond_xf s' \<noteq> 0}" and B="B;; setter"]; clarsimp)
-     apply (frule intermediate_Normal_state[OF _ _ body_inv])
-      apply fastforce
-     apply clarsimp
-     apply (rename_tac inter_t)
-     apply (frule hoarep_exec[OF _ _ body_inv, rotated], fastforce)
-     apply (frule_tac s=inter_t in hoarep_exec[rotated 2], fastforce+)[1]
-    apply (metis (mono_tags, lifting) HoarePartial.SeqSwap empty_iff exec_abrupt mem_Collect_eq)
-   apply (metis (mono_tags, lifting) HoarePartial.SeqSwap exec_stuck mem_Collect_eq)
-  apply (metis (mono_tags, lifting) HoarePartial.SeqSwap empty_iff exec_fault mem_Collect_eq)
-  done
-
 text \<open>
-  This rule is intended to be used to show correspondence between a Haskell whileLoop and a C
-  while loop, where in particular, the C while loop is parsed into a Simpl While loop that
-  updates the cstate as part of the loop condition. In such a case, the CParser will produce a Simpl
-  program in the form that is seen in the conclusion of this rule.\<close>
+  This rule is intended to be used to show correspondence between a @{term whileLoop} and a Simpl
```

Comment:
I could add in another bullet point here talking about the requirement that the `whileLoop` condition be in the `the blah` form and that this cannot fail, and therefore, we must have the `no_ofail` assumption.

I think @lsf37 and I discussed briefly the idea that we might want an alternate `whileLoop` where the condition is always a reader monad function, and where the new `whileLoop` fails if the reader monad function fails. But that would be a lot of mucking around, especially given all the rules we already have about `whileLoop`

Reply:
Yes, adding that bullet point would make sense, with a remark that eliminating the `no_ofail` would require a slightly different form of `whileLoop` that can deal with failure in the condition.